jupyter-docker/delocalizing_contents_manager.py

```diff
@@ -0,0 +1,152 @@
+from datetime import datetime, timedelta
+import json
+import os
+import re
+import subprocess
+from notebook.services.contents.filemanager import FileContentsManager
+
+METADATA_TTL = timedelta(minutes=5)
+
+class DelocalizingContentsManager(FileContentsManager):
+  """
+  A contents manager allowing for configurable automated delocalization.
+
+  Files will always be persisted to the local Jupyter file system as usual. The
+  decision of whether to try delocalization is driven by the presence of a file
+  named .delocalize.json on the Jupyter file system. The format of this
+  JSON is as follows:
+
+  {
+    // Optional, only delocalize files from the collocated directory with a
+    // basename matching this Python "re" regex. If unspecified, all collocated
+    // files will be matched.
+    "pattern": "",
+    // Required. Cloud storage directory in which to persist matching files.
+    "destination": ""
+  }
+
+  gsutil is used to delocalize files. A log is produced on the Jupyter server
+  in the root directory. As a rule, this logfile should only be created if
+  delocalization is configured in at least one subdirectory.
+
+  Note: when a new .delocalize.json config is added, existing eligible files
+  will not be automatically backfilled. They will only be delocalized upon
+  further modification.
+  """
+
+  # TODO: Support GCS version preconditions on save.
+  # TODO: Try shifting the delocalization logic into a background thread.
+
+  # Cache delocalization metadata for 5 minutes. The cache points to a tuple of
+  # (metadata: dict, read_at: datetime), with a key of JSON file path. Negative
+  # lookups are stored here as well with None as metadata (this will be the common
+  # case for any clusters which don't use delocalization).
+  def __init__(self, *args, **kwargs):
+    self.log.info('initializing DelocalizingContentsManager')
+    self.delocalize_metadata = {}
+    # Allows for stubbing in tests.
+    self.file_cmd = ['gsutil', '-q']
```

Comment:
Should we think about specifying `-m` as well?
```
$ gsutil help options

  -m          Causes supported operations (acl ch, acl set, cp, mv, rm, rsync,
              and setmeta) to run in parallel. This can significantly improve
              performance if you are performing operations on a large number of
              files over a reasonably fast network connection.
```